File: Tests/ConnectorControlCoreTests/ClaudeConfigIOTests.swift
import XCTest
import ConnectorControlTestSupport
@testable import ConnectorControlCore

final class ClaudeConfigIOTests: XCTestCase {
    var tempDir: TempDir!
    var dir: URL!
    var url: URL { dir.appendingPathComponent("claude_desktop_config.json") }

    override func setUpWithError() throws {
        tempDir = TempDir(prefix: "claude")
        dir = tempDir.url
    }

    override func tearDownWithError() throws {
        tempDir.dispose()
    }

    private func write(_ s: String) throws { try Data(s.utf8).write(to: url) }

    private func rootObject() throws -> [String: Any] {
        try XCTUnwrap(JSONSerialization.jsonObject(
            with: Data(contentsOf: url)) as? [String: Any])
    }

    func testReadRealisticConfig() throws {
        try write(Fixtures.realisticClaudeConfig)
        let servers = try ClaudeConfigIO.readMCPServers(at: url)
        XCTAssertEqual(Set(servers.keys), ["scoutbook", "aws-mcp", "service-now"])
        XCTAssertEqual(servers["scoutbook"], RemotePattern.make(url: "https://scoutbook.example.com/mcp"))
    }

    func testReadMissingFileReturnsEmpty() throws {
        XCTAssertEqual(try ClaudeConfigIO.readMCPServers(at: url), [:])
    }

    func testReadAbsentKeyReturnsEmpty() throws {
        try write(#"{"preferences": {}}"#)
        XCTAssertEqual(try ClaudeConfigIO.readMCPServers(at: url), [:])
    }

    func testReadMalformedFileThrows() throws {
        try write("{oops")
        XCTAssertThrowsError(try ClaudeConfigIO.readMCPServers(at: url)) {
            guard case ClaudeConfigError.malformed = $0 else {
                return XCTFail("wrong error: \($0)")
            }
        }
    }

    func testReadNonObjectMCPServersThrows() throws {
        try write(#"{"mcpServers": "surprise"}"#)
        XCTAssertThrowsError(try ClaudeConfigIO.readMCPServers(at: url))
    }

    func testWritePreservesEveryOtherKeyByValue() throws {
        try write(Fixtures.realisticClaudeConfig)
        let before = try rootObject()
        try ClaudeConfigIO.write(
            mcpServers: ["only-one": .object(["command": .string("echo")])], to: url)
        let after = try rootObject()
        XCTAssertEqual(Set(after.keys), Set(before.keys))
        for key in before.keys where key != "mcpServers" {
            XCTAssertEqual(
                try JSONSerialization.data(withJSONObject: ["v": after[key]!],
                                           options: [.sortedKeys]),
                try JSONSerialization.data(withJSONObject: ["v": before[key]!],
                                           options: [.sortedKeys]),
                "key \(key) changed")
        }
        let servers = try XCTUnwrap(after["mcpServers"] as? [String: Any])
        XCTAssertEqual(Array(servers.keys), ["only-one"])
    }

    func testWriteToMissingFileCreatesIt() throws {
        try ClaudeConfigIO.write(
            mcpServers: ["a": .object(["command": .string("x")])], to: url)
        XCTAssertEqual(Set(try rootObject().keys), ["mcpServers"])
    }

    func testReadEmptyFileReturnsEmptyServers() throws {
        try write("")
        XCTAssertEqual(try ClaudeConfigIO.readMCPServers(at: url), [:])
    }

    func testWriteToEmptyFileRecreatesConfig() throws {
        try write("")
        try ClaudeConfigIO.write(
            mcpServers: ["a": .object(["command": .string("x")])], to: url)
        XCTAssertEqual(Set(try rootObject().keys), ["mcpServers"])
        XCTAssertEqual(Set(try ClaudeConfigIO.readMCPServers(at: url).keys), ["a"])
    }

    func testWriteRefusesMalformedFile() throws {
        try write("{oops")
        XCTAssertThrowsError(try ClaudeConfigIO.write(mcpServers: [:], to: url))
        XCTAssertEqual(try String(contentsOf: url, encoding: .utf8), "{oops")
    }

    func testDisabledSubsetOmittedAndReadBack() throws {
        try write(Fixtures.realisticClaudeConfig)
        var servers = try ClaudeConfigIO.readMCPServers(at: url)
        servers.removeValue(forKey: "aws-mcp")
        try ClaudeConfigIO.write(mcpServers: servers, to: url)
        XCTAssertEqual(Set(try ClaudeConfigIO.readMCPServers(at: url).keys),
                       ["scoutbook", "service-now"])
    }
}

File: windows/tests/ConnectorControl.Core.Tests/ClaudeConfigIOTests.cs
using ConnectorControl.Core.Tests.TestSupport;

namespace ConnectorControl.Core.Tests;

public class ClaudeConfigIOTests : IDisposable
{
    private readonly TempDir dir = new("claude");
    private string Url => dir.File("claude_desktop_config.json");

    public void Dispose() => dir.Dispose();

    private void Write(string s) => File.WriteAllText(Url, s);

    private JsonValue Root() => JsonValue.Parse(File.ReadAllBytes(Url));

    private static readonly JsonValue Scoutbook = JsonValue.Object(
        ("command", JsonValue.String("npx")),
        ("args", JsonValue.Array([JsonValue.String("-y"), JsonValue.String("mcp-remote"), JsonValue.String("https://scoutbook.example.com/mcp")])));

    [Fact]
    public void ReadRealisticConfig()
    {
        Write(Fixtures.RealisticClaudeConfig);
        var servers = ClaudeConfigIO.ReadMcpServers(Url);
        Assert.Equal(new HashSet<string> { "scoutbook", "aws-mcp", "service-now" }, servers.Keys.ToHashSet());
        Assert.Equal(Scoutbook, servers["scoutbook"]);
    }

    [Fact]
    public void ReadMissingFileReturnsEmpty()
    {
        Assert.Empty(ClaudeConfigIO.ReadMcpServers(Url));
    }

    [Fact]
    public void ReadAbsentKeyReturnsEmpty()
    {
        Write("{\"preferences\": {}}");
        Assert.Empty(ClaudeConfigIO.ReadMcpServers(Url));
    }

    [Fact]
    public void ReadMalformedFileThrows()
    {
        Write("{oops");
        Assert.Throws<ClaudeConfigException>(() => ClaudeConfigIO.ReadMcpServers(Url));
    }

    [Fact]
    public void ReadNonObjectMcpServersThrows()
    {
        Write("{\"mcpServers\": \"surprise\"}");
        var ex = Assert.Throws<ClaudeConfigException>(() => ClaudeConfigIO.ReadMcpServers(Url));
        Assert.Equal("mcpServers is not a JSON object", ex.Detail);
    }

    [Fact]
    public void ReadNonObjectTopLevelThrows()
    {
        Write("[1, 2]");
        var ex = Assert.Throws<ClaudeConfigException>(() => ClaudeConfigIO.ReadMcpServers(Url));
        Assert.Equal("top level is not a JSON object", ex.Detail);
    }

    [Fact]
    public void WritePreservesEveryOtherKeyByValue()
    {
        Write(Fixtures.RealisticClaudeConfig);
        var before = Root();
        ClaudeConfigIO.Write(new Dictionary<string, JsonValue> { ["only-one"] = JsonValue.Object(("command", JsonValue.String("echo"))) }, Url);
        var after = Root();
        Assert.Equal(before.ObjectProperties.Keys.ToHashSet(), after.ObjectProperties.Keys.ToHashSet());
        foreach (var key in before.ObjectProperties.Keys.Where(k => k != "mcpServers"))
        {
            Assert.Equal(before[key], after[key]);
        }
        Assert.Equal(["only-one"], after["mcpServers"]!.ObjectProperties.Keys.ToArray());
    }

    [Fact]
    public void WriteToMissingFileCreatesIt()
    {
        ClaudeConfigIO.Write(new Dictionary<string, JsonValue> { ["a"] = JsonValue.Object(("command", JsonValue.String("x"))) }, Url);
        Assert.Equal(["mcpServers"], Root().ObjectProperties.Keys.ToArray());
    }

    [Fact]
    public void ReadEmptyFileReturnsEmptyServers()
    {
        Write("");
        Assert.Empty(ClaudeConfigIO.ReadMcpServers(Url));
    }

    [Fact]
    public void WriteToEmptyFileRecreatesConfig()
    {
        Write("");
        ClaudeConfigIO.Write(new Dictionary<string, JsonValue> { ["a"] = JsonValue.Object(("command", JsonValue.String("x"))) }, Url);
        Assert.Equal(["mcpServers"], Root().ObjectProperties.Keys.ToArray());
        Assert.Equal(["a"], ClaudeConfigIO.ReadMcpServers(Url).Keys.ToArray());
    }

    [Fact]
    public void WriteRefusesMalformedFile()
    {
        Write("{oops");
        Assert.Throws<ClaudeConfigException>(() => ClaudeConfigIO.Write(new Dictionary<string, JsonValue>(), Url));
        Assert.Equal("{oops", File.ReadAllText(Url));
    }

    [Fact]
    public void DisabledSubsetOmittedAndReadBack()
    {
        Write(Fixtures.RealisticClaudeConfig);
        var servers = new Dictionary<string, JsonValue>(ClaudeConfigIO.ReadMcpServers(Url));
        servers.Remove("aws-mcp");
        ClaudeConfigIO.Write(servers, Url);
        Assert.Equal(new HashSet<string> { "scoutbook", "service-now" }, ClaudeConfigIO.ReadMcpServers(Url).Keys.ToHashSet());
    }

    [Fact]
    public void WriteUsesAppleSerializationFormat()
    {
        Write("{\"preferences\": {\"sidebarMode\": \"x\"}, \"Zeta\": 1, \"alpha\": 0.5}");
        ClaudeConfigIO.Write(new Dictionary<string, JsonValue> { ["s"] = JsonValue.Object(("command", JsonValue.String("npx"))) }, Url);
        const string expected =
            "{\n  \"alpha\" : 0.5,\n  \"mcpServers\" : {\n    \"s\" : {\n      \"command\" : \"npx\"\n    }\n  },\n  \"preferences\" : {\n    \"sidebarMode\" : \"x\"\n  },\n  \"Zeta\" : 1\n}";
        Assert.Equal(expected, File.ReadAllText(Url));
    }
}
